File: app/src/main/java/me/rei_m/hyakuninisshu/store/MaterialStore.java
/*
 * Copyright (c) 2017. Rei Matsushita
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is
 * distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See
 * the License for the specific language governing permissions and limitations under the License.
 */

package me.rei_m.hyakuninisshu.store;

import android.support.annotation.NonNull;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import javax.inject.Inject;
import javax.inject.Singleton;

import io.reactivex.Observable;
import io.reactivex.subjects.BehaviorSubject;
import me.rei_m.hyakuninisshu.action.Dispatcher;
import me.rei_m.hyakuninisshu.action.karuta.EditKarutaAction;
import me.rei_m.hyakuninisshu.action.material.FetchMaterialAction;
import me.rei_m.hyakuninisshu.domain.model.karuta.Karuta;
import me.rei_m.hyakuninisshu.domain.model.karuta.KarutaIdentifier;

@Singleton
public class MaterialStore {

    private final BehaviorSubject<List<Karuta>> karutaListSubject = BehaviorSubject.create();
    public final Observable<List<Karuta>> karutaList = karutaListSubject;

    public final Observable<Karuta> karuta(@NonNull KarutaIdentifier karutaId) {
        return karutaListSubject.flatMap(value -> Observable.fromIterable(value).filter(karuta -> karuta.identifier().equals(karutaId)));
    }

    @Inject
    public MaterialStore(@NonNull Dispatcher dispatcher) {
        dispatcher.on(FetchMaterialAction.class).subscribe(action -> {
            karutaListSubject.onNext(action.karutas.asList());
        });
        dispatcher.on(EditKarutaAction.class).subscribe(action -> {
            if (karutaListSubject.hasValue()) {
                List<Karuta> karutaList = new ArrayList<>(karutaListSubject.getValue());
                karutaList.set(karutaList.indexOf(action.karuta), action.karuta);
                karutaListSubject.onNext(Collections.unmodifiableList(karutaList));
            }
        });
    }
}

File: state/src/main/java/me/rei_m/hyakuninisshu/state/material/store/MaterialStore.kt
/*
 * Copyright (c) 2020. Rei Matsushita
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is
 * distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See
 * the License for the specific language governing permissions and limitations under the License.
 */

package me.rei_m.hyakuninisshu.state.material.store

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import me.rei_m.hyakuninisshu.state.core.Dispatcher
import me.rei_m.hyakuninisshu.state.core.Store
import me.rei_m.hyakuninisshu.state.material.action.EditMaterialAction
import me.rei_m.hyakuninisshu.state.material.action.FetchMaterialListAction
import me.rei_m.hyakuninisshu.state.material.model.Material
import javax.inject.Inject

/**
 * 資料の状態を管理する.
 */
class MaterialStore @Inject constructor(dispatcher: Dispatcher) : Store() {

    private val _materialList = MutableLiveData<List<Material>?>()
    val materialList: LiveData<List<Material>?> = _materialList

    private val _isFailure = MutableLiveData(false)
    val isFailure: LiveData<Boolean> = _isFailure

    init {
        register(dispatcher.on(FetchMaterialListAction::class.java).subscribe {
            when (it) {
                is FetchMaterialListAction.Success -> {
                    _materialList.value = it.materialList
                    _isFailure.value = false
                }
                is FetchMaterialListAction.Failure -> {
                    _isFailure.value = true
                }
            }
        }, dispatcher.on(EditMaterialAction::class.java).subscribe {
            when (it) {
                is EditMaterialAction.Success -> {
                    _isFailure.value = false
                    val currentValue = materialList.value ?: return@subscribe
                    _materialList.value = currentValue.map { m ->
                        if (m.no == it.material.no) it.material else m
                    }
                }
                is EditMaterialAction.Failure -> {
                    _isFailure.value = true
                }
            }
        })
    }
}
